cpp/src/holtwinters/internal/hw_utils.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cuda_runtime.h>
+#include <iostream>
+#include <vector>
+#include "../holtwinters_params.h"
+#include "common/cumlHandle.hpp"
+#include "common/cuml_allocator.hpp"
+#include "common/device_buffer.hpp"
+#include "linalg/cublas_wrappers.h"
+#include "linalg/cusolver_wrappers.h"
+#include "linalg/eltwise.h"
+#include "utils.h"
+
+#define IDX(n, m, N) (n + (m) * (N))
+
+#define STMP_EPS (1e-6)
+
+#define GOLD \
+  0.381966011250105151795413165634361882279690820194237137864551377294739537181097550292792795810608862515245
+#define PG_EPS 1e-10
+
+#define SUBSTITUTE(a, b, c, d) \
+  (a) = (b);                   \
+  (c) = (d);
+
+#define MAX_BLOCKS_PER_DIM 65535
+
+#define GET_TID (blockIdx.x * blockDim.x + threadIdx.x)
+
+inline int GET_THREADS_PER_BLOCK(const int n, const int max_threads = 512) {
+  int ret;
+  if (n <= 128)
+    ret = 32;
+  else if (n <= 1024)
+    ret = 128;
+  else
+    ret = 512;
+  return ret > max_threads ? max_threads : ret;
+}
+
+inline int GET_NUM_BLOCKS(const int n, const int max_threads = 512,
+                          const int max_blocks = MAX_BLOCKS_PER_DIM) {
+  int ret = (n - 1) / GET_THREADS_PER_BLOCK(n, max_threads) + 1;
+  return ret > max_blocks ? max_blocks : ret;
+}
+
+template <typename Dtype>
```

Comment:
How will we account for Dtype? `float` will call `fminf`, and `double` will call `fmin`

Reply:
One way to implement this, which might be a little cleaner, is to just have a single function that calls the proper math function underneath based on the precision. For instance:

```
template<typename Dtype>
__device__ Dtype abs_device(Dtype val) {
    int nbytes = sizeof(val);
    ASSERT(nbytes == sizeof(float) || nbytes == sizeof(double), "Dtype needs to be single or double precision floating point");
    if(nbytes == sizeof(float)) return fabsf(val);
    else return fabs(val);
}
```

This is possible because the template is assumed to be a floating point type, so you route the type explicitly here based on its size. `sizeof()` is also compile-time so there should be no overhead in doing it this way. It also makes the intent much more direct, instead of the developer having to rely on extra template functions being compiled in.